go-controller/pkg/ovn/logical_switch_cache.go

```diff
@@ -0,0 +1,165 @@
+package ovn
+
+import (
+	"fmt"
+	"net"
+	"reflect"
+	"sync"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/config"
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/util"
+	"k8s.io/klog"
+	ipam "k8s.io/kubernetes/pkg/registry/core/service/ipallocator"
+)
+
+// logicalSwitchInfo contains information corresponding to the node. It holds the
+// subnet allocations (v4 and v6) as well as the IPAM allocator instances for each
+// subnet managed for this node
+// logicalSwitchInfo is used to save switch specific information such as subnets,
+// allocated IPs per switch etc.
+type logicalSwitchInfo struct {
+	hostSubnets []*net.IPNet
+	ipams       []ipam.Interface
+}
+
+type ipamFactoryFunc func(*net.IPNet) ipam.Interface
+
+type logicalSwitchCache struct {
+	cache map[string]logicalSwitchInfo
+	// A RW mutex for logicalSwitchCache which holds logicalSwitch information
+	sync.RWMutex
+	ipamFunc ipamFactoryFunc
+}
+
+func NewIPAMAllocator(cidr *net.IPNet) ipam.Interface {
+	return ipam.NewCIDRRange(cidr)
+}
+
+func (cache *logicalSwitchCache) Add(nodeName string, hostSubnets []*net.IPNet) error {
+	cache.Lock()
+	defer cache.Unlock()
+	if existing, ok := cache.cache[nodeName]; ok && !reflect.DeepEqual(existing.hostSubnets, hostSubnets) {
+		klog.Warningf("Node %q logical switch already in cache with subnet %s; replacing with %s", nodeName,
+			util.JoinIPNets(existing.hostSubnets, ","), util.JoinIPNets(hostSubnets, ","))
+	}
+	var ipams []ipam.Interface
+	for _, subnet := range hostSubnets {
+		ipam := cache.ipamFunc(subnet)
+		if ipam == nil {
+			klog.Warningf("IPAM for subnet %s was not initialized for node %q", subnet, nodeName)
+			continue
+		}
+		// reserve IPs for gw, management port, hybrid overlay etc.
+		if err := reserveIps(subnet, ipam); err != nil {
+			klog.Warningf("Failed reserving IPs: %s", err)
+			continue
+		}
+		ipams = append(ipams, ipam)
+	}
+	cache.cache[nodeName] = logicalSwitchInfo{
+		hostSubnets: hostSubnets,
+		ipams:       ipams}
+
+	return nil
+}
+
+func (cache *logicalSwitchCache) Delete(nodeName string) error {
+	cache.Lock()
```

Comment:
I think this function need not return `error` as there is no error case to return

Reply:
Ack